src/main/java/io/appium/java_client/DefaultGenericMobileDriver.java

```diff
@@ -157,7 +157,8 @@ public List findElementsByXPath(String using) {
 
     @Override
     public String toString() {
+        Capabilities capabilities = getCapabilities();
         return String.format("%s, Capabilities: %s", getClass().getCanonicalName(),
-                getCapabilities().asMap().toString());
+                capabilities != null ? capabilities.asMap().toString() : "null");
```

Comment:
I'd rather use checkNotNull macros everywhere

Reply:
or this is normal that `getCapabilities` can sometimes return null?

Reply:
`getCapabilities` will be null **until** the response of the *new session* request, where the real capabilities are returned by the server.

As I mentioned in the description, if the *new session* command fails (which can be pretty often when you are trying to set up the proper version combination of Appium, Automation and device), this `NullPointerException` masks the real failure (with **no** exception nesting).

So, overall, this makes the initial set up incredibly frustrating.

Reply:
yep, I've checked the Selenium code and it's how it currently works. Although I would like to get the response to https://github.com/SeleniumHQ/selenium/issues/6903. What if this is just a bug and not the expected behaviour?

Reply:
I'm ok about learning the *official posture*.  However, if `RemoteWebDriver` does indeed have a typo (missing a `this.`), then the exception masking would still remain.  But instead of `NullPointerException`, a `ClassCastException` would be thrown:

https://github.com/appium/java-client/blob/fa71401de3fdb74eef7e4e7d513cf0c8e6adbe36/src/main/java/io/appium/java_client/android/AndroidDriver.java#L210

In such a case, the fix would be much uglier.